tests/framework/extensions/codecoverage/codecoverage.go

```diff
@@ -0,0 +1,181 @@
+
+func retrieveCodeCoverageFile(client *rancher.Client, clusterID, coverageFilename string, podNames []string) error {
+	kubeConfig, err := kubeconfig.GetKubeconfig(client, clusterID)
+	if err != nil {
+		return err
+	}
+
+	restConfig, err := (*kubeConfig).ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	for _, podName := range podNames {
+		fileName := fmt.Sprintf("%s%s", podName, coverageFilename)
+		dst := fmt.Sprintf("%s/%s", outputDir, fileName)
+
+		err := kubeconfig.CopyFileFromPod(restConfig, podName, cattleSystemNameSpace, coverageFilename, dst)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// KillRancherTestServicesRetrieveCoverage is a function that kills the rancher service of the local cluster
+// inorder for the code coverage report to be written, and then copies over the coverage reports from the pods
+// to a local destination. The custom code coverage rancher image must be running in the local cluster.
+func KillRancherTestServicesRetrieveCoverage(client *rancher.Client) error {
+	var podNames []string
+	dynamicClient, err := client.GetRancherDynamicClient()
+	if err != nil {
+		return err
+	}
+
+	pods, err := dynamicClient.Resource(corev1.SchemeGroupVersion.WithResource("pods")).Namespace(cattleSystemNameSpace).List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, pod := range pods.Items {
+		name := pod.GetName()
+		if strings.Contains(name, "rancher") && !strings.Contains(name, "webhook") {
+			podNames = append(podNames, pod.GetName())
+		}
+	}
+
+	err = killTestServices(client, localCluster, podNames)
+	if err != nil {
+		return err
+	}
+
+	err = checkServiceIsRunning(dynamicClient)
+	if err != nil {
+		return err
+	}
+
+	return retrieveCodeCoverageFile(client, localCluster, rancherCoverFile, podNames)
+}
+
+// KillAgentTestServicesRetrieveCoverage is a function that kills the cattle-cluster-agent service of a downstream cluster
+// inorder for the code coverage report to be written, and then copies over the coverage reports from the pods
+// to a local destination. The custom code coverage rancher-agent image must be running in the downstream cluster.
+func KillAgentTestServicesRetrieveCoverage(client *rancher.Client) error {
+	clusters, err := client.Steve.SteveType(clusters.ProvisioningSteveResouceType).ListAll(&types.ListOpts{})
+	if err != nil {
+		return err
+	}
+
+	for _, cluster := range clusters.Data {
+		clusterStatus := &apiv1.ClusterStatus{}
+		err = v1.ConvertToK8sType(cluster.Status, clusterStatus)
+		if err != nil {
+			return err
+		}
+		clusterID := clusterStatus.ClusterName
+		if clusterID != localCluster {
+			dynamicClient, err := client.GetDownStreamClusterClient(clusterID)
+			if err != nil {
+				logrus.Errorf("could not connect to downstream cluster")
+				continue
+			}
+
+			pods, err := dynamicClient.Resource(corev1.SchemeGroupVersion.WithResource("pods")).Namespace("cattle-system").List(context.Background(), metav1.ListOptions{})
```

Comment:
This is for the cattle-cluster-agent service though, right? So wouldn't it need to be here as well since on line 102, that is for the local cluster and not a downstream cluster like this is doing.

Reply:
What do you mean Markus? The cattle-cluster-agent pods are in the cattle-system namespace as well. I'm not quite following.

Reply:
Ignore my last comment, please, I was confusing myself 😄